dbms/src/Interpreters/Join.cpp

```diff
@@ -2138,9 +2496,415 @@ BlockInputStreamPtr Join::createStreamWithNonJoinedRows(const Block & left_sampl
+    return hashToSelector(hash);
+}
+
+void Join::spillMostMemoryUsedPartitionIfNeed()
+{
+    Int64 target_partition_index = -1;
+    size_t max_bytes = 0;
+    Blocks blocks_to_spill;
+
+    {
+        std::unique_lock lk(build_probe_mutex);
+#ifdef DBMS_PUBLIC_GTEST
+        // for join spill to disk gtest
+        if (restore_round == 1 && spilled_partition_indexes.size() >= partitions.size() / 2)
+            return;
+#endif
+        if (max_bytes_before_external_join && getTotalByteCount() <= max_bytes_before_external_join)
+        {
+            return;
+        }
+        for (size_t j = 0; j < partitions.size(); ++j)
+        {
+            if (!partitions[j]->spill && (target_partition_index == -1 || partitions[j]->memory_usage > max_bytes))
+            {
+                target_partition_index = j;
+                max_bytes = partitions[j]->memory_usage;
+            }
+        }
+        if (target_partition_index == -1)
+        {
+            return;
+        }
+
+        RUNTIME_CHECK_MSG(restore_round < 4, "max_bytes_before_external_join is too small, join restore round exceeds limit, please make max_bytes_before_external_join larger and try again.");
+        RUNTIME_CHECK_MSG(build_concurrency > 1, "Join with build concurrency = 1 does not support spill.");
+        is_spilled = true;
+
+        LOG_DEBUG(log, fmt::format("all bytes used : {}", getTotalByteCount()));
+        for (size_t index = 0; index < partitions.size(); ++index)
+        {
+            LOG_DEBUG(log, fmt::format("partition id : {}, rows : {}, bytes : {}", index, partitions[index]->build_partition.rows, partitions[index]->memory_usage));
+        }
+        LOG_DEBUG(log, fmt::format("make round : {}, partition : {} spill.", restore_round, target_partition_index));
+
+        std::unique_lock partition_lock(partitions[target_partition_index]->partition_mutex);
+        partitions[target_partition_index]->spill = true;
+        releaseBuildPartitionHashTable(target_partition_index, partition_lock);
+        spilled_partition_indexes.push_back(target_partition_index);
+        blocks_to_spill = trySpillBuildPartition(target_partition_index, true, partition_lock);
+    }
+    FAIL_POINT_TRIGGER_EXCEPTION(FailPoints::random_join_spill_to_disk_failpoint);
+    build_spiller->spillBlocks(blocks_to_spill, target_partition_index);
+    LOG_DEBUG(log, fmt::format("all bytes used after spill : {}", getTotalByteCount()));
+}
+
+void Join::insertBlockToBuildPartition(Block && block, size_t partition_index)
+{
+    size_t rows = block.rows();
+    size_t bytes = block.bytes();
+    partitions[partition_index]->build_partition.rows += rows;
+    partitions[partition_index]->build_partition.bytes += bytes;
+    partitions[partition_index]->build_partition.blocks.push_back(block);
+    partitions[partition_index]->build_partition.original_blocks.push_back(std::move(block));
+    partitions[partition_index]->memory_usage += bytes;
```

Comment:
Wrap into `Partition` and just call `partitions[partition_index]->insertBlockForBuild`?

Reply:
> Wrap into `Partition` and just call `partitions[partition_index]->insertBlockForBuild`?

done.